tests/unit_tests/test_analytics/test_sentry.py

```diff
@@ -0,0 +1,387 @@
+    key=INTERNAL_SENTRY_DSN_KEY,
+    address=API_SERVER_ADDRESS,
+    project=INTERNAL_SENTRY_PROJECT,
+)
+EXTERNAL_DSN_URL_FORMAT = SENTRY_DSN_FORMAT.format(
+    key=EXTERNAL_SENTRY_DSN_KEY,
+    address=API_SERVER_ADDRESS,
+    project=EXTERNAL_SENTRY_PROJECT,
+)
+
+
+@pytest.fixture(scope="module", autouse=True)
+def relay():
+    relay_server = MetricRelayServer()
+    relay_server.start()
+
+    return relay_server
+
+
+@pytest.fixture(scope="module", autouse=True)
+def setup_env(relay):
+    os.environ[wandb.env.ERROR_REPORTING] = "True"
+    os.environ[wandb.env.SENTRY_DSN] = WANDB_DSN_URL_FORMAT.format(port=relay.port)
+
+
+class MetricRelayServer:
+    events = {}
+
+    def __init__(
+        self,
+    ) -> None:
+        self.is_running = False
+        self.port = self._get_free_port()
+        self.app = flask.Flask(__name__)
+        self.app.add_url_rule(
+            rule="/api/<project_id>/envelope/",
+            methods=["POST"],
+            view_func=self.sentry,
+        )
+        self.app.add_url_rule(
+            rule="/ping/",
+            methods=["GET"],
+            view_func=self.ping,
+        )
+        self.relay_url = f"http://127.0.0.1:{self.port}"
+
+        self._sentry_responses = []
+
+    def start(self) -> None:
+        # run server in a separate thread
+        relay_server_thread = threading.Thread(
+            target=self.app.run,
+            kwargs={
+                "port": self.port,
+            },
+            daemon=True,
+        )
+        relay_server_thread.start()
+
+        # wait for server to start
+        while not self.is_running:
+            try:
+                requests.get(f"{self.relay_url}/ping/")
+                self.is_running = True
+            except BaseException:
+                continue
+
+        self.is_running = True
+
+    @staticmethod
+    def _get_free_port() -> int:
+        sock = socket.socket()
+        sock.bind(("", 0))
+        _, port = sock.getsockname()
+        return port
+
+    def extract_sentry_message_details(self, event_data):
+        decompressed_data = zlib.decompress(event_data, 16 + zlib.MAX_WBITS)
+        envelope = Envelope.deserialize(decompressed_data)
+        return envelope
+
+    def sentry(self, project_id):
+        decompressed_data = zlib.decompress(request.get_data(), 16 + zlib.MAX_WBITS)
+        envelope = Envelope.deserialize(decompressed_data)
+        self.events[envelope.headers["event_id"]] = {
+            "envelope": envelope,
+            "project_id": project_id,
+        }
+        return "OK"
+
+    def ping(self):
+        return "OK"
+
+
+def wait_for_events(relay, event_ids, timeout=5):
+    start_time = time.time()
+    end_time = start_time + timeout
+
+    while (
+        not all(event_id in relay.events for event_id in event_ids)
+        and time.time() < end_time
+    ):
+        time.sleep(0.1)
+
+    for event_id in event_ids:
+        assert event_id in relay.events
```

Comment:
this sleep is really not ideal for unit testing (or in general in testing)
i'm pretty sure it will cause quite a bit of flake

Reply:
I agree, the issue is that sentry_sdk appears to be sending events asynchronously. Without waiting for some time the tests consistently fail. 

This method mitigates flakiness by having a default 5 second timeout. Because the events are sent over localhost. There is no real latency, just waiting for sentry_sdk to actually send the events.

Reply:
yeah, i get the reasoning behind it, just thinking maybe this approach is not optimal. 
did you check how sentry does it testing internally? ideally if we could mock a bunch of things and assert on that, that would be preferable

Reply:
An option is to implement our own [transport class](https://github.com/getsentry/sentry-python/blob/master/tests/test_monitor.py#L23-L26) and capture the events as they are called from there.

We would then need to add the ability to provide a transport class to the wandb Sentry class.

Reply:
Thanks, i will need to think about it.